diff --git a/libheif/plugins/encoder_kvazaar.cc b/libheif/plugins/encoder_kvazaar.cc
index 7affa6d2..a3fbc388 100644
--- a/libheif/plugins/encoder_kvazaar.cc
+++ b/libheif/plugins/encoder_kvazaar.cc
@@ -338,7 +338,13 @@ static void kvazaar_query_input_colorspace(heif_colorspace* colorspace, heif_chr
   }
   else {
     *colorspace = heif_colorspace_YCbCr;
+#if !defined(KVZ_RANGE_EXTENSION)
     *chroma = heif_chroma_420;
+#else
+    if (*chroma != heif_chroma_420 && *chroma != heif_chroma_422 && *chroma != heif_chroma_444) {
+      *chroma = heif_chroma_420;
+    }
+#endif
   }
 }
 
@@ -351,10 +357,16 @@ static void kvazaar_query_input_colorspace2(void* encoder_raw, heif_colorspace*
   }
   else {
     *colorspace = heif_colorspace_YCbCr;
+#if !defined(KVZ_RANGE_EXTENSION)
     if (*chroma != heif_chroma_420) {
       // Encoding to 4:2:2 and 4:4:4 currently does not work with Kvazaar (https://github.com/ultravideo/kvazaar/issues/418).
       *chroma = heif_chroma_420;
     }
+#else
+    if (*chroma != heif_chroma_420 && *chroma != heif_chroma_422 && *chroma != heif_chroma_444) {
+      *chroma = heif_chroma_420;
+    }
+#endif
   }
 }
 
@@ -525,17 +537,34 @@ static heif_error kvazaar_start_sequence_encoding_intern(void* encoder_raw, cons
   config->framerate_num = framerate_num;
   config->framerate_denom = framerate_denom;
 
+  const char* format_str = "P420";
   if (isGreyscale) {
+    format_str = "P400";
     config->input_format = KVZ_FORMAT_P400;
+    config->chroma_format = KVZ_CSP_400;
+    config->chroma_shift_w = 0;
+    config->chroma_shift_h = 0;
   }
   else if (chroma == heif_chroma_420) {
+    format_str = "P420";
     config->input_format = KVZ_FORMAT_P420;
+    config->chroma_format = KVZ_CSP_420;
+    config->chroma_shift_w = 1;
+    config->chroma_shift_h = 1;
   }
   else if (chroma == heif_chroma_422) {
+    format_str = "P422";
     config->input_format = KVZ_FORMAT_P422;
+    config->chroma_format = KVZ_CSP_422;
+    config->chroma_shift_w = 1;
+    config->chroma_shift_h = 0;
   }
   else if (chroma == heif_chroma_444) {
+    format_str = "P444";
     config->input_format = KVZ_FORMAT_P444;
+    config->chroma_format = KVZ_CSP_444;
+    config->chroma_shift_w = 0;
+    config->chroma_shift_h = 0;
   }
   else {
     return heif_error{
@@ -545,6 +574,8 @@ static heif_error kvazaar_start_sequence_encoding_intern(void* encoder_raw, cons
     };
   }
 
+  api->config_parse(config, "input-format", format_str);
+
 
   heif_color_profile_nclx* nclx = nullptr;
   heif_error err = heif_image_get_nclx_color_profile(image, &nclx);
