diff --git a/libheif/plugins/encoder_kvazaar.cc b/libheif/plugins/encoder_kvazaar.cc
index 7affa6d2..8c711ec8 100644
--- a/libheif/plugins/encoder_kvazaar.cc
+++ b/libheif/plugins/encoder_kvazaar.cc
@@ -338,7 +338,13 @@ static void kvazaar_query_input_colorspace(heif_colorspace* colorspace, heif_chr
   }
   else {
     *colorspace = heif_colorspace_YCbCr;
+#if !defined(KVZ_RANGE_EXTENSION)
     *chroma = heif_chroma_420;
+#else
+    if (*chroma != heif_chroma_420 && *chroma != heif_chroma_422 && *chroma != heif_chroma_444) {
+      *chroma = heif_chroma_420;
+    }
+#endif
   }
 }
 
@@ -351,10 +357,16 @@ static void kvazaar_query_input_colorspace2(void* encoder_raw, heif_colorspace*
   }
   else {
     *colorspace = heif_colorspace_YCbCr;
+#if !defined(KVZ_RANGE_EXTENSION)
     if (*chroma != heif_chroma_420) {
       // Encoding to 4:2:2 and 4:4:4 currently does not work with Kvazaar (https://github.com/ultravideo/kvazaar/issues/418).
       *chroma = heif_chroma_420;
     }
+#else
+    if (*chroma != heif_chroma_420 && *chroma != heif_chroma_422 && *chroma != heif_chroma_444) {
+      *chroma = heif_chroma_420;
+    }
+#endif
   }
 }
 
